lib/ClangImporter/ClangImporter.cpp

```diff
@@ -7471,6 +7471,80 @@ bool importer::isForeignReferenceTypeWithoutImmortalAttrs(const clang::QualType
          !hasImmortalAtts(pointeeType->getDecl());
 }
 
+// Returns the given declaration along with all its parent declarations that are
+// reference types.
+llvm::SmallVector<const clang::RecordDecl *, 4>
+getRefParentDecls(const clang::RecordDecl *decl) {
+  assert(decl && "decl is null inside getRefParentDecls");
+
+  llvm::SmallVector<const clang::RecordDecl *, 4> matchingDecls;
+
+  if (hasImportAsRefAttr(decl)) {
+    matchingDecls.push_back(decl);
+  }
+
+  if (const auto *cxxRecordDecl = llvm::dyn_cast<clang::CXXRecordDecl>(decl)) {
+    if (!cxxRecordDecl->hasDefinition())
+      return matchingDecls;
+    cxxRecordDecl->forallBases([&](const clang::CXXRecordDecl *baseDecl) {
+      if (hasImportAsRefAttr(baseDecl)) {
+        matchingDecls.push_back(baseDecl);
+      }
+      return true;
+    });
+  }
+
+  return matchingDecls;
+}
+
+const clang::RecordDecl *getRefParentOrDiag(
+    const clang::RecordDecl *decl,
+    std::optional<std::reference_wrapper<ClangImporter::Implementation>>
+        importerImpl) {
+  auto refParentDecls = getRefParentDecls(decl);
+  if (refParentDecls.empty()) {
+    return nullptr;
+  }
+
+  llvm::StringRef uniqueRetainAttr, uniqueReleaseAttr;
+
+  for (const auto *refParentDecl : refParentDecls) {
+    assert(refParentDecl && "refParentDecl is null inside getRefParentOrDiag");
+    for (const auto *attr : refParentDecl->getAttrs()) {
+      if (const auto swiftAttr = llvm::dyn_cast<clang::SwiftAttrAttr>(attr)) {
+        const auto &attribute = swiftAttr->getAttribute();
+        if (attribute.starts_with("retain:")) {
+          if (uniqueRetainAttr.empty()) {
+            uniqueRetainAttr = attribute;
+          } else if (uniqueRetainAttr != attribute) {
+            uniqueRetainAttr = "";
+            break;
+          }
+        } else if (attribute.starts_with("release:")) {
+          if (uniqueReleaseAttr.empty()) {
+            uniqueReleaseAttr = attribute;
+          } else if (uniqueReleaseAttr != attribute) {
+            uniqueReleaseAttr = "";
+            break;
+          }
+        }
+      }
+    }
+  }
+
+  // Ensure exactly one retain and one release attribute
```

Comment:
I think we do not check yet whether we have multiple overloads for the retain/release functions with the same name.